kv/tikv/storage/commands/rollback.go

```diff
@@ -0,0 +1,95 @@
+package commands
+
+import (
+	"fmt"
+	"github.com/pingcap-incubator/tinykv/kv/tikv/storage/kvstore"
+	"github.com/pingcap-incubator/tinykv/proto/pkg/kvrpcpb"
+)
+
+type Rollback struct {
+	request *kvrpcpb.BatchRollbackRequest
+}
+
+func NewRollback(request *kvrpcpb.BatchRollbackRequest) Rollback {
+	return Rollback{request}
+}
+
+func (r *Rollback) BuildTxn(txn *kvstore.MvccTxn) error {
+	txn.StartTS = &r.request.StartVersion
+	for _, k := range r.request.Keys {
+		err := rollbackKey(k, txn)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func rollbackKey(key []byte, txn *kvstore.MvccTxn) error {
+	lock, err := txn.GetLock(key)
+	if err != nil {
+		return err
+	}
+
+	if lock == nil || lock.TS != *txn.StartTS {
+		// There is no lock, check the write status.
+		existingWrite, err := txn.FindWrite(key, *txn.StartTS)
+		if err != nil {
+			return err
+		}
+		if existingWrite == nil {
+			// There is no write either, presumably the prewrite was lost. We insert a rollback write anyway.
+			write := kvstore.Write{StartTS: *txn.StartTS, Kind: kvstore.WriteKindRollback}
+			txn.PutWrite(key, &write, *txn.StartTS)
+
+			return nil
+		} else {
+			if existingWrite.Kind == kvstore.WriteKindRollback {
+				// The key has already been rolled back, so nothing to do.
+				return nil
+			}
+
+			// The key has already been committed. This should not happen since the client should never send both
+			// commit and rollback requests.
+			return fmt.Errorf("storage: rollback: key has already been committed %v", key)
```

Comment:
add a predefined error?

Reply:
This is an error that should never happen and if it happens in real life should end up in a panic somewhere, rather than an error which is possible and is sent back to the client. Given that, I thought it was better practice in Go to use a string error?